### PR TITLE
Separate off-ledger ond on-ledger speedy

### PR DESCRIPTION
**Pull request by cocreature**
changelog_begin
changelog_end

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala [hunk 1]
```diff
@@ -97,8 +97,9 @@ private[lf] object Speedy {
 
   private type Actuals = util.ArrayList[SValue]
 
-  /** The speedy CEK machine. */
-  final class Machine(
+  private[lf] sealed trait LedgerMode
```
Comment (cocreature) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala:100]: Very open to better naming suggestions.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala [hunk 2]
```diff
@@ -156,14 +163,20 @@ private[lf] object Speedy {
...
-      val onLedger: Boolean,
+      val ledgerMode: LedgerMode,
   ) {
 
     /* kont manipulation... */
 
     @inline
     private[speedy] def kontDepth(): Int = kontStack.size()
 
+    private[lf] def withOnLedger[T](op: String)(f: OnLedger => T): T =
```
Comment (cocreature) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala:174]: Not quite sure if this callback-style approach makes that much sense here. The alternative would be a `getOnLedger` or something like that.

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/SBuiltinTest.scala [hunk 3]
```diff
@@ -1413,7 +1415,18 @@ object SBuiltinTest {
   }
 
   private def evalSExpr(e: SExpr, onLedger: Boolean): Either[SError, SValue] = {
-    val machine = Speedy.Machine.fromPureSExpr(compiledPackages, e, onLedger)
+    val machine = if (onLedger) {
```
Comment (cocreature) [inline daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/SBuiltinTest.scala:1418]: This seems a bit messy. Not quite sure what would be better. Maybe just a helper method to evaluate `pure x` in a scenario context?

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala [hunk 4]
```diff
@@ -840,7 +841,7 @@ private[lf] object SBuiltin {
     override private[speedy] final def execute(
         args: util.ArrayList[SValue],
         machine: Machine,
-    ): Unit = {
+    ): Unit = machine.withOnLedger("SBUCreate") { onLedger =>
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:844]: Why not defining a an abstract class for `OnLedgerBuiltin` ?  Something like :+1: 
```scala
sealed abstract class OnLedgerBuiltin(arity: Int) extends SBuiltin(arity) with Product {
    
    protected def execute(
        args: util.ArrayList[SValue],
        machine: Machine,
        onLedger: OnLedger
    ): Unit

    final override def execute(args: util.ArrayList[SValue], machine: Machine): Unit =
      machine.withOnLedger(productPrefix)(execute(args, machine, _))
  }
```

Comment (cocreature) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:844]: sounds good, let me give it a try

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala [hunk 5]
```diff
@@ -1250,18 +1252,21 @@ private[lf] object SBuiltin {
       }
     }
 
-    private[this] def executeCommit(args: util.ArrayList[SValue], machine: Machine): Unit =
-      machine.ptx
+    private[this] def executeCommit(
+        args: util.ArrayList[SValue],
+        machine: Machine,
+        onLedger: OnLedger): Unit =
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:1258]:
```suggestion
        onLedger: OnLedger,
    ): Unit =
```

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/SBuiltinTest.scala (hunk 3 above)
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/SBuiltinTest.scala:1418]: I think it is OK for tests.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala (hunk 1 above)
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala:100]: I do not like it neither, but I do not have better alternative. :disappointed:

**Review by remyhaemmerle-da — APPROVED** [review]
Very nice. Thanks for pushing this through.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala (hunk 4 above)
Comment (cocreature) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala]: Added.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala (hunk 5 above)
Comment (cocreature) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:1283]: You really love your trailing commas :slightly_smiling_face:

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala (hunk 4 above)
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala]: :+1:

File: daml-lf/scenario-interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/ScenarioRunner.scala [hunk 6]
```diff
@@ -29,6 +29,7 @@ final case class ScenarioRunner(
     machine: Speedy.Machine,
     partyNameMangler: (String => String) = identity) {
   var ledger: ScenarioLedger = ScenarioLedger.initialLedger(Time.Timestamp.Epoch)
+  val onLedger = machine.withOnLedger("ScenarioRunner")(identity)
```
Comment (aherrmann-da) [inline daml-lf/scenario-interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/ScenarioRunner.scala:32]: This feels a bit hacky. Maybe that's a bias coming from Haskell, but to me `withOnLedger` implies that the `onLedger` should be limited to the scope of the continuation and should not leak out. Either `machine.getOnLedger` or moving `machine.withOnLedger` to the relevant places would seem cleaner.

File: daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/LedgerInteraction.scala [hunk 7]
```diff
@@ -388,6 +388,7 @@ class IdeClient(val compiledPackages: CompiledPackages) extends ScriptLedgerClie
     outputTransactionVersions = transaction.TransactionVersions.DevOutputVersions,
     traceLog = traceLog,
   )
+  val onLedger = machine.withOnLedger("IdeClient")(identity)
```
Comment (aherrmann-da) [inline daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/LedgerInteraction.scala:391]: Same here.

**Review by aherrmann-da — APPROVED** [review]
LGTM, thank you!

Not so sure about the `withOnLedger(identity)` pattern though.

File: daml-lf/scenario-interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/ScenarioRunner.scala (hunk 6 above)
Comment (cocreature) [reply daml-lf/scenario-interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/ScenarioRunner.scala:32]: Yeah I agree that it’s a bit weird. The reason why I went for this is that the machine is t the class level so it seems weird to limit the scope of the onledger stuff to anything less.

Comment (cocreature) [reply daml-lf/scenario-interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/ScenarioRunner.scala]: I’ve inlined the implementation in those two cases to make this a bit less confusing.

Comment (aherrmann-da) [reply daml-lf/scenario-interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/ScenarioRunner.scala]: Thanks, yes, that's clearer.

**Comment by remyhaemmerle-da** [comment]
fixes #7387

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala (hunk 1 above)
Comment (hurryabit) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala:100]: What do you think about something like `ExecutionEnvironment`? You can either execute DAML on a ledger or off a ledger.

Comment (cocreature) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala:100]: sounds reasonable but also sounds very generic to the point where it doesn’t give you much of an idea of what the possible execution environments could be.

**Pull request merged**